src/Microsoft.Identity.Client/UI/CustomWebUiHandler.cs

```diff
@@ -0,0 +1,95 @@
+﻿// ------------------------------------------------------------------------------
+// 
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+// 
+// This code is licensed under the MIT License.
+// 
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+// 
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+// 
+// ------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Extensibility;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.OAuth2;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.UI
+{
+    internal class CustomWebUiHandler : IWebUI
```

Comment:
Wouldn't it be clearner if instead of ICustomWebUI we would simply expose IWebUI? This way, all our IWebUI implementations and the users's customs web ui are interchangable.  It shouldn't be difficult to achieve, just need to refactor IWebUI AcquireAuthorizationAsync to take 2 Uris and return a Task<Uri> instead of Task<AuthorizationResult>

Reply:
and would we drop the request context?

Reply:
Yes, we should drop the RequestContext and the AuthorizationResult. Implementers need to return only the URI containing the auth code. They are responsible for their own logging.

Reply:
I disagree.  ICustomWebUI is a public API that needs stability and for us to not change it.  IWebUI is internal and gives us the flexibility to change it, and also enables us to log in our internal flows since we have RequestContext.  If we drop RequestContext then anything in our internal code won't have access to log in this path, and we do quite a bit of logging for cancel/failure/success in this path.  But my main reason for wanting to keep these separate is internal API flexibility and not to have a public API be our primary gateway for internals since that reduces our flexibility.  We also need different handling of the return from how some of the browser flows handle the URI response and we nicely encapsulate that in our CustomUiWebHandler class.

Reply:
Extensiblity: We already have 7 or 8 implementations of IWebUI, which is a simple interface. In VS we used to have the principle that "everything is extensible" and we would try to use the same extensiblity points that we gave to users.

RequestContext: it can be passed in via the constructor of the objects implementing IWebUI.

I didn't understand your last sentence - that we need custom handling of the return value. I was under the impression that all current implementations of IWebUI execute some logic to extract an uri that contains the auth code and then construct the AuthorizationResult object. We can move this construction elsewhere, i.e. remove a responsability from IWebUI

Reply:
That all seems reasonable, Bogdan.  If the high level approach/implementation as it is works, I'd like to get a drop of this to VS for them to test and get this concern off the table with them.  Then as part of finishing this work we can do the refactor of IWebUI to this model before merging.  I don't want to do that additional work before letting VS try this out to ensure it will unblock their scenario.  I'll hold this PR until we've completed the feature and I think you've convinced me we can use IWebUI everywhere once we have buyoff from VS that the basic plumbing works for them.  That way if we get feedback from them we can incorporate before finalizing this review.